File: other/tonetable/make_tone_table.pl
#!/usr/bin/env perl
# Input: A data file where each row is like:
#   啊\t123\ta1
# Output: A HTML table of hanzi grouped by tone.

use strict;
use warnings;

# Get the data from the file.
open(DATA, "frequency_pinyin_table.txt");
my %SYLLABLES = ();
while (my $record = <DATA>) {
    next if ($record =~ /^\s*$/ || $record =~ /^#/);
    my ($zi, $rank, $pinyin) = ($record =~ m/^(\S+)\s(\d+)\s(\S+)/);
    $rank = int($rank);
    my $is_primary = !($pinyin =~ m/[*]$/);
    my $freq = frequency_level($rank);
    my $is_frequent = ($freq < 6);
    if ($is_primary and $is_frequent) {
        my ($syll, $tone) = ($pinyin =~ m/^(.+)(\d)$/);
        $tone = int($tone);
        $tone = 0 if $tone == 5;
        $SYLLABLES{$syll}[$tone][$freq] .= $zi;
    }
}
close(DATA);

# Assign a level of a character based on it's rank in an ordered list
# From most to least frequent.
sub frequency_level {
    my $rank = shift;
    return 0 if ($rank < 250);
    return 1 if ($rank < 500);
    return 2 if ($rank < 1000);
    return 3 if ($rank < 1500);
    return 4 if ($rank < 2000);
    return 5 if ($rank < 2500);
    return 6;
}

# Generate and output the html.
sub print_row;
sub join_initial_final;
my @initials = split /\s/,
    "- b p m f d t n l g k h z c s zh ch sh r j q x";
my @finals = split /\s/,
    "a o e ai ei ao ou an en ang eng ong
    i ia iao ie iou ian in iang ing iong
    u ua uo uai ui uan uen uang ueng v ve van vn";
print "<table>\n";
print "<tr><th></th><th>0</th><th>1</th><th>2</th><th>3</th><th>4</th></tr>\n";
print_row "er";
for my $final (@finals) {
    for my $initial (@initials) {
        my $syllable = join_initial_final($initial, $final);
        my $has_entry = $SYLLABLES{$syllable};
        print_row $syllable if $has_entry;
    }
}
print "</table>";

# Given one scalar syllable, print suitable html tr element.
sub print_row {
    my $syllable = shift;
    print "<tr><th class='syllable'>$syllable</th>\n";
    for my $tone (0..4) {
        if (!$SYLLABLES{$syllable}[$tone]) {
            print "\t<td class='tone$tone' class='empty'></td>\n";
        } else {
            print "\t<td class='tone$tone'>";
            for my $freq (0..5) {
                my $hanzi = $SYLLABLES{$syllable}[$tone][$freq];
                print "<span class='freq$freq'>$hanzi</span>" if $hanzi;
            }
            print "</td>\n";
        }
    }
    print "</tr>\n";
}

# Given initial and final, return the standard way of writing in pinyin.
sub join_initial_final {
    my ($init, $fin) = @_;
    if ($init =~ m/-/) {
        $fin =~ s/^u$/wu/;
        $fin =~ s/^i$/yi/;
        $fin =~ s/^ui$/wei/;
        $fin =~ s/^v/yu/;
        $fin =~ s/^i/y/;
        $fin =~ s/^u/w/;
        $init = "";
    } else {
        $fin =~ s/^iou/iu/;
        $fin =~ s/^uen/un/;
    }
    if ($init =~ m/[jqx]/) {
        $fin =~ s/^u/X/;
        $fin =~ s/v/u/;
    }
    return $init . $fin;
}

File: chinese/tonetable/make_tone_table.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Generate a HTML tone table from frequency_pinyin_table.txt
Mimics the behavior of make_tone_table.pl
"""

import re
from collections import defaultdict


def frequency_level(rank):
    """Assign a frequency level based on rank."""
    rank = int(rank)
    if rank < 250:
        return 0
    elif rank < 500:
        return 1
    elif rank < 1000:
        return 2
    elif rank < 1500:
        return 3
    elif rank < 2000:
        return 4
    elif rank < 2500:
        return 5
    else:
        return 6


def join_initial_final(initial, final):
    """Given initial and final, return the standard way of writing in pinyin."""
    if initial == "-":
        if final == "u":
            final = "wu"
        elif final == "i":
            final = "yi"
        elif final == "ui":
            final = "wei"
        elif final.startswith("v"):
            final = "yu" + final[1:]
        elif final.startswith("i"):
            final = "y" + final[1:]
        elif final.startswith("u"):
            final = "w" + final[1:]
        initial = ""
    else:
        if final == "iou":
            final = "iu"
        elif final == "uen":
            final = "un"

    if initial in ["j", "q", "x"]:
        if final.startswith("u"):
            final = "X" + final[1:]
        final = final.replace("v", "u")

    return initial + final


def parse_data_file(filename):
    """Parse the frequency_pinyin_table.txt file."""
    syllables = defaultdict(lambda: defaultdict(lambda: defaultdict(str)))

    with open(filename, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue

            # Parse: char\trank\tpinyin
            match = re.match(r"^(\S+)\s+(\d+)\s+(\S+)$", line)
            if not match:
                continue

            zi, rank, pinyin = match.groups()
            rank = int(rank)
            is_primary = not pinyin.endswith("*")
            freq = frequency_level(rank)
            is_frequent = freq < 6

            if is_primary and is_frequent:
                # Extract syllable and tone
                tone_match = re.match(r"^(.+?)(\d)$", pinyin)
                if tone_match:
                    syll, tone = tone_match.groups()
                    tone = int(tone)
                    if tone == 5:  # Convert tone 5 to 0 (neutral tone)
                        tone = 0
                    syllables[syll][tone][freq] += zi

    return syllables


def generate_html_table(syllables):
    """Generate the HTML table."""
    # Define the syllable patterns to generate
    initials = [
        "-",
        "b",
        "p",
        "m",
        "f",
        "d",
        "t",
        "n",
        "l",
        "g",
        "k",
        "h",
        "z",
        "c",
        "s",
        "zh",
        "ch",
        "sh",
        "r",
        "j",
        "q",
        "x",
    ]

    finals = [
        "a",
        "o",
        "e",
        "ai",
        "ei",
        "ao",
        "ou",
        "an",
        "en",
        "ang",
        "eng",
        "ong",
        "i",
        "ia",
        "iao",
        "ie",
        "iou",
        "ian",
        "in",
        "iang",
        "ing",
        "iong",
        "u",
        "ua",
        "uo",
        "uai",
        "ui",
        "uan",
        "uen",
        "uang",
        "ueng",
        "v",
        "ve",
        "van",
        "vn",
    ]

    html_lines = []
    html_lines.append("<table>")
    html_lines.append(
        "<tr><th></th><th>0</th><th>1</th><th>2</th><th>3</th><th>4</th></tr>"
    )

    # Print "er" first
    print_row("er", syllables, html_lines)

    # Print other syllables
    for final in finals:
        for initial in initials:
            syllable = join_initial_final(initial, final)
            if syllable in syllables:
                print_row(syllable, syllables, html_lines)

    html_lines.append("</table>")
    return "\n".join(html_lines)


def print_row(syllable, syllables, html_lines):
    """Print a table row for a syllable."""
    html_lines.append(f"<tr><th class='syllable'>{syllable}</th>")

    for tone in range(5):  # 0-4
        if not syllables[syllable][tone]:
            html_lines.append(f"\t<td class='tone{tone}' class='empty'></td>")
        else:
            html_lines.append(f"\t<td class='tone{tone}'>")
            for freq in range(6):  # 0-5
                hanzi = syllables[syllable][tone][freq]
                if hanzi:
                    html_lines.append(f"<span class='freq{freq}'>{hanzi}</span>")
            html_lines.append("</td>")

    html_lines.append("</tr>")


def main():
    """Main function."""
    input_file = "frequency_pinyin_table.txt"
    output_file = "tone_table.html"

    print(f"Reading data from {input_file}...")
    syllables = parse_data_file(input_file)

    print("Generating HTML table...")
    html_table = generate_html_table(syllables)

    print(f"Writing HTML to {output_file}...")
    with open(output_file, "w", encoding="utf-8") as f:
        f.write(html_table)

    print("Done!")


if __name__ == "__main__":
    main()
